stdlib/src/collections/list.mojo

```diff
@@ -581,13 +580,16 @@ struct List[T: CollectionElement](CollectionElement, Sized, Boolable):
         Args:
             new_size: The new size.
         """
-        debug_assert(
-            new_size <= self.size,
-            (
-                "New size must be smaller than or equal to current size when no"
-                " new value is provided."
-            ),
-        )
+        if self.size < new_size:
```

Comment:
Somehow this makes me feel like we should have an `always_assert` or something similar that is not dependent on the build type. That's the only issue with `debug_assert` here, right?

Reply:
If we do that, the string is going to be materialised unconditionally, which is not good, right?

Reply:
Could you explain why variadic can help mitigate this problem?

Reply:
> I wouldn't worry too much about `debug_assert` as I think it will be dead code eliminated in the release, right?

Yes, but it still poses a problem if you want a debug build on a target that cannot heap allocate.

If you pass a variadic of Stringables, the implementation of the function can ensure that they only get actually stringified (and thefore for needing a heap allocation) if the passed condition fails. So instead of
```mojo
my_assert(cond, "some message with id" + str(id))
```
you would have
```mojo
my_assert(cond, "some message with id", id)
```
There might be control flow sinking happening in the first case, but it's not guaranteed, while in the second case an implementation like this would make sure that no heap allocation happens if the condition is true:
```mojo
fn my_assert[*Ts: Stringable](cond: Bool, *msgs: *Ts):
    if not cond:
        @parameter
        if sys.can_heap_allocate():  # doesn't exist, but you get the point
            msg: String = ""
            for x in msgs:  # might have to fiddle with the exact syntax here
                msg += str(x)  # many reallocs, but who cares if you are already aborting?
            abort(msg)
        else:
            alias static_msg = "something went wrong"
            abort(static_msg)
```